src/gateway.rs

```diff
@@ -173,18 +182,30 @@ impl Gateway {
         }
     }
 
-    async fn handle_transmit_beacon(&mut self, logger: &Logger, beacon: Beacon) {
+    fn tx_power(&mut self, logger: &Logger) -> Option<u32> {
         let region_params = if let Some(region_params) = &self.region_params {
             region_params
         } else {
             warn!(logger, "ignoring transmit request, no region params");
```

Comment:
I'd prefer tx_power just return a Result<u32> instead of an Option<u32> and not log in something that's just there to return a value? i.e. do the logging in the caller(s)?

Reply:
is it considered "correct" if there's no tx_power? either way we're sort of bailing out on sending the beacon in the caller so seems like you could just as easily have the final value returned as a result?

Reply:
My point was that the tx_power function is logging on behalf of the caller and ends up logging, for example, about "beacon" even when it's being used for a downlink.

Reply:
Two parts to that statement; I agree I wouldn’t have this function do the logging but to the other part of your comment I’m wondering if it’s appropriate for this function to return an option or a result

Reply:
We are logging from a `fn(&self)` - seems innocuous to me; error msg does need fixing though.

If we make this a result type, each `tx_power` caller (`handle_beacon`, `handle_downlink`) will have to unpack Result, draft an error message, etc. Seems tedious when the point of refactoring this into a function was to reduce the repetition between those two functions.

Reply:
right but at least they _can_ provide contextual logging